### PR TITLE
8043251: Bogus javac error: required: no arguments, found: no arguments

### PR DESCRIPTION
**Pull request by archiecobbs**
We have an error message `compiler.err.cant.apply.symbol` for cases where a method invocation doesn't resolve. It shows the "required" parameters and the "found" parameters, plus a further description of the problem.

This message being used inappropriately when the problem is actually due to explicitly passed method type parameters instead of regular method parameters.

For example, if you do this:
```java
Function<String, String> f = Function.<String, String>identity();
```
the error reported is this:
```
error: method identity in interface Function<T,R> cannot be applied to given types;
    Function<String, String> f = Function.<String, String>identity();
                                         ^
  required: no arguments
  found:    no arguments
  reason: wrong number of type arguments; required 1
  where T,R are type-variables:
    T extends Object declared in interface Function
    R extends Object declared in interface Function
```
The real error here is `wrong number of type arguments; required 1`, but the `compiler.err.cant.apply.symbol` error message that assumes the problem is with the regular parameters, which it displays. The result is the appearance of the useless and confusing `required` and `found` lines.

This patch creates an alternate version of that error message (`compiler.err.cant.apply.symbol.noargs`) that omits the `required` and `found` lines, and makes the compiler use this alternate message when the error is due to type parameters (specifically, when the underlying error is "wrong number of type arguments" or "explicit type argument X does not conform to declared bound(s)").

This improves the error for example above to this:
```
error: method identity in interface Function<T,R> cannot be applied to given types;
    Function<String, String> f = Function.<String, String>identity();
                                         ^
  reason: wrong number of type arguments; required 1
  where T,R are type-variables:
    T extends Object declared in interface Function
    R extends Object declared in interface Function
1 error
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8043251](https://bugs.openjdk.org/browse/JDK-8043251): Bogus javac error: required: no arguments, found: no arguments

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [92338830](https://git.openjdk.org/jdk/pull/10799/files/923388304cbab7f89fe820f0c3406229acf6da61)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10799/head:pull/10799` \
`$ git checkout pull/10799`

Update a local copy of the PR: \
`$ git checkout pull/10799` \
`$ git pull https://git.openjdk.org/jdk pull/10799/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10799`

View PR using the GUI difftool: \
`$ git pr show -t 10799`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10799.diff">https://git.openjdk.org/jdk/pull/10799.diff</a>

</details>

### CODE REVIEW

**Comment by archiecobbs** [comment]
keepalive

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java [hunk 1]
```diff
@@ -4173,17 +4173,34 @@ JCDiagnostic getDiagnostic(JCDiagnostic.DiagnosticType dkind,
...
-                      methodArguments(ws.type.getParameterTypes()),
-                      methodArguments(argtypes),
-                      kindName(ws.owner),
-                      ws.owner.type,
-                      c.snd);
+
+            // If the problem is due to type arguments, then the method parameters aren't relevant,
+            // so use the error message that omits them to avoid confusion.
+            switch (c.snd.getCode()) {
+            case "compiler.misc.wrong.number.type.args":
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:4180]: minor indentation suggestion: I would move the content of the `switch` to the right

Comment (archiecobbs) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java]: Thanks, it is inconsistent with the style in that file. Fixed in 3f8c9509861.

**Review by vicente-romero-oracle — APPROVED** [review]
looks sensible

**Pull request closed without merging**